### DISCUSSION TITLE
Is it safe to expose the name of a route?

### DISCUSSION THREAD

**Discussion by alexisLefebvre** [symfony/symfony#47197] [Q&A] [answered]
We use a custom controller to switch between several domains and languages, in order to do that we pass the name of the route and the country as parameters of the URL.

Is it safe? In worst case, someone could use this to generate random routes, but they will be redirected to it and then be blocked by the firewall, so it looks like it won't be an issue, but maybe we are missing something.

**Comment by RSickenberg** [comment, marked as answer]
In my opinion, there are no risks whatsoever. 
In the meantime, why do you not pass the full translated URI instead?

**Reply by alexisLefebvre** [reply]
Thanks for your answer. We have a complex setup with routes with or without language prefix, depending on domains and countries, some countries have 2 languages so URIs are different.

**Reply by lyrixx** [reply]
I still dont understand. On one app we did that:

```twig
{% for lang in locales|filter(lang => lang != app.request.locale and) %}
    <li role="option">
        <a href="{{ translatedurl(app.request.attributes.get('_route'), app.request.attributes.get('_route_params')|merge({ _locale: lang })|merge(app.request.query.all), lang) }}"
           class="c-language__link">
            {{ imageMacros.lazy_img(asset('img/flag-' ~ lang ~ '.svg'), ('width="24" height="16" alt="' ~(('language.' ~ lang)|trans) ~ '"'), amp) }}
            <span class="u-pl-sm">{{ ('language.' ~ lang)|trans }}</span>
        </a>
    </li>
{% endfor %}

```

**Reply by alexisLefebvre** [reply]
@lyrixx Could you please explain what is the `translatedurl` function?

We used a custom Twig function that rode the routing files, but the processing was too heavy and took 400 ms, probably because it didn't used cached resources. So instead we now pass the route name and generate the URI only when the user want to switch the country. I don't remember exactly our problem, but it was related to the fact that Belgium must support `fr` and `nl` locales, so we add them as prefix, while it's not the case for `fr` on the `fr` domain. Hence we can't map routes easily.

**Reply by lyrixx** [reply]
Nothing very clever here. But we have some translated content in the DB. here is our code:

```php
    public function translatedUrl($routeName, $routeParams, $locale, bool $returnNullOnInvalid = false): string
    {
        $router = $this->container->get('router');
        $manager = $this->container->get(PageManager::class);
        $doctrine = $this->container->get('doctrine');

        $routeParams['_locale'] = $locale;

        switch ($routeName) {
            case 'city_amp':
            case 'city_show':
                $city = $manager->getCity($routeParams['slug']);
                $routeParams['slug'] = $this->translationObject($locale, $city, 'slug');
                break;
            case 'product_amp':
            case 'product_show':
                $product = $manager->getProduct($routeParams['slug']);
                $routeParams['slug'] = $this->translationObject($locale, $product, 'slug');
                break;
            case 'cms':
                $currentLocale = $this->container->get('request_stack')->getCurrentRequest()->getLocale();
                $slug = $doctrine->getRepository(Page::class)->getAlternativeLocaleSlug($routeParams['slug'], $currentLocale, $locale);

                if (null === $slug) {
                    $routeParams = [];
                    $routeName = 'home_show';
                } else {
                    $routeParams['slug'] = $slug;
                }
        }

        try {
            $route = $router->generate($routeName, $routeParams, UrlGeneratorInterface::ABSOLUTE_URL);
        } catch (RouteNotFoundException $e) {
            if ($returnNullOnInvalid) {
                return null;
            }
            $route = $router->generate('home_show', $routeParams, UrlGeneratorInterface::ABSOLUTE_URL);
        }

        return $route;
    }
```

**Reply by alexisLefebvre** [reply]
Thanks for your help, we have to dig more in our code, we know that it's too complicated but we didn't found a better solution and it's not one of our priorities.
